code/modules/reagents/chemistry/holder.dm

```diff
@@ -100,19 +109,29 @@
  * * list/data - Any reagent data for this reagent, used for transferring data with reagents
  * * reagtemp - Temperature of this reagent, will be equalized
  * * no_react - prevents reactions being triggered by this addition
+ * * added_purity - override to force a purity when added
+ * * added_pH - override to force a pH when added
+ * * override_base_pH - bypass the pH update when adding a reagent, causing it to retain the pH of the solution it's being added to
  */
-/datum/reagents/proc/add_reagent(reagent, amount, list/data=null, reagtemp = 300, no_react = 0)
+/datum/reagents/proc/add_reagent(reagent, amount, list/data=null, reagtemp = 300, added_purity = INFINITY, added_pH, no_react = 0, override_base_pH = FALSE)
 	if(!isnum(amount) || !amount)
 		return FALSE
 
-	if(amount <= 0)
+	if(amount <= CHEMICAL_QUANTISATION_LEVEL)//To prevent small amount problems.
 		return FALSE
 
 	var/datum/reagent/glob_reagent = GLOB.chemical_reagents_list[reagent]
 	if(!glob_reagent)
 		WARNING("[my_atom] attempted to add a reagent called '[reagent]' which doesn't exist. ([usr])")
 		return FALSE
 
+	var/datum/reagent/D = GLOB.chemical_reagents_list[reagent]
+	if(added_purity == INFINITY) //Because purity additions can be 0
```

Comment:
If you are using `INFINITY` as a key for not adding purity, you'd be better served by a non-numeric value, `null`, which you can check with `isnull(value)`.

Reply:
I always thought null was 0 in byond! I'm glad to hear it's not.